designs/2019-relative-plugin-loading/README.md

```diff
@@ -0,0 +1,169 @@
+* ESLint loads plugins from the CWD of the process, which is typically somewhere in the end user's project. However, with the current setup the correct place to load plugins from is the ESLint invoker's project. As a result, if the ESLint invoker is not the end user (e.g. in `create-react-app` projects), the default plugin-loading location is incorrect. The ESLint invoker can change the CWD, but changing the CWD has some inconvenient side-effects:
+    * Only the `node_modules` folder in the CWD is ignored by default, so changing the CWD would cause some files to be unignored.
+    * Editor integrations usually won't know to change the CWD (since they might not be able to figure out who the "intended" ESLint invoker is for a given project), so they might fail to load the correct plugins.
+* Users often want to write custom project-specific rules and put them in the project's repository. However, plugins currently need to be in `node_modules`, because allowing other paths for plugins would introduce the possibility of name conflicts.
+
+## Detailed Design
+
+The design has three components:
+
+* Plugins are resolved relative to the config files that depend on them, rather than being loaded relative to the ESLint invoker's project.
+* Plugin name conflicts result in a fatal error.
+* Plugins in configs can be loaded by filepath.
+
+### Plugins are resolved relative to the config files that depend on them
+
+A config is said to "load" a given plugin name if either of the following are true:
+
+* The config includes the plugin name in a `plugins` list, e.g. `plugins: ["foo"]`
+* The config extends a shareable config from the plugin, e.g. `extends: ["plugin:foo/bar"]`
+
+To load a plugin from a config, the following steps are taken:
+
+* If the config loading the name `eslint-plugin-X` is itself a bundled config in plugin called `eslint-plugin-X`, then the load resolves to that plugin itself. (This is called a *self-load*.)
+* Otherwise, the plugin is loaded as a Node module relative to the location of the config file that loads the plugin's name.
+
+**Open question:** Should `extends: ["plugin:foo/bar"]` be considered a "load" of a plugin? (The "Open Questions" section at near the bottom of this RFC contains more details on how this would affect behavior. To understand the tradeoffs, it may be easier to read through this RFC first before considering the open question.)
+
+### Plugin name conflicts result in a fatal error
+
+If at least two different config files load the same plugin name, excluding self-loads, then a fatal config error is thrown.
```

Comment:
I'm considering if we can check package.json.

- When it found a plugin in a config file, it reads package.json of the package which contains the config file. If the `dependencies` field doesn't include the plugin, it excludes the config file from conflict check.

This means, the following cases will be ignored in conflict check naturally:

- shareable configs which use `peerDependencies` for plugins
- plugin configs which loads itself
- `.eslintrc` files which use plugins that shareable configs imported (as expecting the package manager hoists the plugins)

Reply:
Hmm, checking `package.json` doesn't seem appealing to me because then we really would be duplicating the behavior of a package manager. I think there are lots of other cases (e.g. `optionalDependencies`, local packages within other packages, etc.) that we would need to deal with -- it seems like it would work in common cases but it could break due to unrelated features of package managers.

I think the cases are worth considering separately:

* To simulate shareable configs that have `peerDependencies` (where multiple configs can use the same version of the same plugin):

    I think any solution here needs to be opt-in for shareable configs. Otherwise, a shareable config could start depending on a different version of the plugin and accidentally break its users.

    But if shareable configs are opting in anyway, then maybe they could just add `eslint-plugin-foo` as a `peerDependency`, remove `plugins: ["foo"]` from their config, and require the user to supply `plugins: ["foo"]`. Then the user would be supplying the `plugins: ["foo"]` directive directly.

> plugin configs which loads itself

I think we can just keep the existing special case here -- it adds a small amount of complexity, but doesn't really cause major problems.

> `.eslintrc` files which use plugins that shareable configs imported (as expecting the package manager hoists the plugins)

I think this would already be adequately solved by using option (2) from https://github.com/eslint/rfcs/pull/14#discussion_r260934656.

Reply:
I don't want to touch the details of a package manager, too. But, I thought touching the minimum spec of `package.json` (rather than a package manager) to relax conflict errors may be nice because the confliction matter causes by the change of `package.json`.

I want to keep the current behavior that we can use `extends:"plugin:foo/bar"` without `plugins` field.